rfcs/0000-buildpack-resource.md

```diff
@@ -0,0 +1,104 @@
+
+The ClusterStore being the single location for all buildpack images used in multiple builders makes managing the available buildpacks within kpack cumbersome. Adding new buildpacks requires modifying an existing resource and removing buildpacks requires carefully selecting the buildpack image to remove from the list of buildpackage images. The `kp` cli was built to handle this complexity but, the `kp` cli should not be a prerequisite to managing buildpacks within kpack.
+
+The ClusterStore being a monolithic specification of all available buildpacks leads to performance issues within kpack. If the list of available buildpacks within a ClusterStore is lengthy the reconciliation of a single ClusterStore can take considerable time. If a single buildpack within the ClusterStore is unavailable reconciliation will fail which will cause the entire contents of the ClusterStore to be unavailable.
+
+**Outcome:**
+
+Allow Buildpacks within kpack to be managed independently of other buildpacks with their own unique lifecycles with a new ClusterBuildpack/Buildpack resource while eventually removing support for the ClusterStore.
+
+**Actions to take:**
+
+Introduce a new ClusterBuildpack and Buildpack resource that provides a buildpack to be available within a Cluster or a singular namespace and be used in ClusterBuilders and Builders.
+
+The Buildpack resource will allow for an individual buildpack to be configured with a single image reference.
+
+Sample Cluster Buildpack Resource
+
+```yaml 
+apiVersion: kpack.io/v1alpha3
+kind: ClusterBuildpack
+metadata:
+  name: paketo-java-buildpack-5.9.1
+spec:
+  image: gcr.io/paketo-buildpacks/java:5.9.1
+``` 
+
+Sample Buildpack Resource
+
+```yaml
+apiVersion: kpack.io/v1alpha3
+kind: Buildpack
+metadata:
+  name: my-custom-app-buildpack-0.3.1
+spec:
+  image: gcr.io/my-org/my-custom-app-buildpack:0.3.1
+``` 
+
+With the Buildpack resource, the ClusterStore resource will not be necessary and with the api that introduces the Buildpack Resource, the ClusterStore object reference should be removed on Builder resources.
+
+An example Builder without a configured ClusterStore is shown below.
+
+```yaml
+apiVersion: kpack.io/v1alpha3
+kind: Builder
+metadata:
+  name: my-builder
+spec:
+  tag: gcr.io/sample/builder
+  stack:
+    name: bionic-stack
+    kind: ClusterStack
+  order:
+  - group:
+    - id: paketo-buildpacks/java      
+```
+The `paketo-buildpacks/java` buildpack referenced in the Builder order would be sourced from the Buildpack resource within the `my-builder` namespace or the ClusterBuildpack resource that provides the `paketo-buildpacks/java` with the highest version number as defined by semver. If multiple Buildpack resources provide the same selected Buildpack the Builder should reconcile with an error.
```

Comment:
Could this get you into a state where someone else creates a buildpack or clusterbuildpack providing the same id and version that would inadvertently cause your builder to error out

Reply:
>If multiple Buildpack resources provide the same selected Buildpack the Builder should reconcile with an error.
 
Maybe buildpacks should fail to reconcile if a buildpack with the same version/id combo already exists.

Reply:
I think this is a risk but, inherent if a builder is letting its versions float without direct references to a buildpack.

Reply:
> Maybe buildpacks should fail to reconcile if a buildpack with the same version/id combo already exists.

I think that would be difficult to enforce.